File: src/main/java/com/atsuishio/superbwarfare/capability/energy/ItemEnergyStorage.java
package com.atsuishio.superbwarfare.capability.energy;

import net.minecraft.world.item.ItemStack;

import java.util.function.Function;

public class ItemEnergyStorage extends DynamicEnergyStorage {

    private static final String NBT_ENERGY = "Energy";

    private final ItemStack stack;

    public ItemEnergyStorage(ItemStack stack, long capacity) {
        this(stack, capacity, capacity, capacity);
    }

    public ItemEnergyStorage(ItemStack stack, long capacity, long maxInsert, long maxExtract) {
        this(stack, s -> capacity, s -> maxInsert, s -> maxExtract);
    }

    public ItemEnergyStorage(ItemStack stack, Function<ItemStack, Long> capacityGetter, Function<ItemStack, Long> maxInsertGetter, Function<ItemStack, Long> maxExtractGetter) {
        super(() -> capacityGetter.apply(stack), () -> maxInsertGetter.apply(stack), () -> maxExtractGetter.apply(stack));

        this.stack = stack;
        if (stack.getTag() != null) {
            this.amount = stack.hasTag() && stack.getTag().contains(NBT_ENERGY) ? stack.getTag().getLong(NBT_ENERGY) : 0;
        }
    }

    @Override
    protected void onFinalCommit() {
        stack.getOrCreateTag().putLong(NBT_ENERGY, getAmount());
    }
}

File: src/main/kotlin/com/atsuishio/superbwarfare/capability/energy/ItemEnergyStorage.kt
package com.atsuishio.superbwarfare.capability.energy

import com.atsuishio.superbwarfare.init.ModDataComponents
import net.minecraft.world.item.ItemStack
import team.reborn.energy.api.EnergyStorage
import java.util.function.Function

class ItemEnergyStorage(
    private val stack: ItemStack,
    capacityGetter: Function<ItemStack, Int>,
    maxReceiveGetter: Function<ItemStack, Int>,
    maxExtractGetter: Function<ItemStack, Int>
) : DynamicEnergyStorage(
    { capacityGetter.apply(stack) },
    { maxReceiveGetter.apply(stack) },
    { maxExtractGetter.apply(stack) }
) {

    @JvmOverloads
    constructor(
        stack: ItemStack,
        capacity: Int,
        maxReceive: Int = capacity,
        maxExtract: Int = capacity
    ) : this(
        stack,
        { _ -> capacity },
        { _ -> maxReceive },
        { _ -> maxExtract }
    )

    init {
        val current = stack.get(EnergyStorage.ENERGY_COMPONENT)
        val legacy = stack.get(ModDataComponents.ENERGY)
        this.energy = current ?: legacy?.toLong() ?: 0L

        if (current == null && legacy != null) {
            stack.set(EnergyStorage.ENERGY_COMPONENT, energy)
            stack.remove(ModDataComponents.ENERGY)
        }
    }

    override fun onFinalCommit() {
        stack.set(EnergyStorage.ENERGY_COMPONENT, amount)
        stack.remove(ModDataComponents.ENERGY)
    }
}
